### DISCUSSION TITLE
Handshake before middleware

### DISCUSSION THREAD

**Discussion by ilyaDegtyarenko** [nuxt/nuxt#17466] [Questions]
Where can I make a handshake request before enabling middlewares? With the placement of the token in cookie for further inquiries.

In the server middleware or plugin, which is better?
If I trying to do this in the app.vue then there it is done after processing the middleware and redirects are performed

I also noticed that if I do this in the plugin, then new tokens are not placed in the cookies and I do not have a new token on the page in useCookie()

**Comment by pi0** [comment]
Hi. If you need something happens before SSR process being started, I would recommend `server/middleware/cookie.ts`

**Reply by ilyaDegtyarenko** [reply]
I need to get user data and set tokens in cookies. User data needs to be installed in the storage (pinia), but there is no access to the storage in server middleware, I will find that you can transfer some data by adding it to the event context, and then get it in a regular middleware and install there, but this is not a good. 
What would you advise?

**Reply by pi0** [reply]
Then you can use a Nuxt plugin. It is the earliest possible stage of SSR that has access to both request event and vue app context to extend anything such as the state. I would use server middleware, only if you need to have a logic that happens on each navigation such as redirection.

> I also noticed that if I do this in the plugin, then new tokens are not placed in the cookies and I do not have a new token on the page in useCookie()

Do you mind sharing your code as a reproduction?

**Reply by ilyaDegtyarenko** [reply]
Yes of course

**_~/server/middleware/handshake.ts_**
```
export default defineEventHandler(async (event: CompatibilityEvent) => {
    ...

    if (!validate(tokens.token)) {
        await refreshToken()
        ...
    } else {
        await handshake()
    }

    ...
    setCookie(event, 'token', tokens.token)
    setCookie(event, 'refreshToken', tokens.refreshToken)

    event.context.handshakePayload = handshakePayload as HandshakePayload
})
```
**_plugins/handshakeHandlerPlugin.server.ts_**
```
export default defineNuxtPlugin(app => {
    const handshakePayload = app.ssrContext.event.context.handshakePayload as HandshakePayload

    if (handshakePayload.error) {
        throwError(handshakePayload.error)
    } else if (handshakePayload.user) {
        useUserStore().$patch({
            user: handshakePayload.user,
            ...
        })
    }
})
```

**Reply by ilyaDegtyarenko** [reply]
For such a case, it is not very suitable, as I understand it.
Where can I request another server to retrieve data that needs to be stored?
